### PR TITLE
add support for multiple powers per quest

### PR DESCRIPTION
**Pull request by AndrewCLu**
- [x] Allow a quest to have multiple powers
- [x] Display quest as number of powers completed
- [x] Display progress bar for PowerCard as number of cards collected

### CODE REVIEW

File: apps/jubmoji-quest/src/pages/index.tsx [hunk 1]
```diff
@@ -61,35 +61,27 @@ export default function Home() {
...
-                const collectionCardIndices = collectionCards.map(
-                  (card) => card.index
-                );
-
                 const questImagePath = imageLink;
 
-                const collected = jubmojis.filter((jubmoji) =>
-                  collectionCardIndices.includes(jubmoji.pubKeyIndex)
+                const numPowersCompleted = powers.filter((power) =>
+                  power.collectionCards
```
Comment (vb7401) [inline apps/jubmoji-quest/src/pages/index.tsx:70]: Don't we need to figure out what type of power this is first before this check? Like if it's NUnique, aren't we supposed to be doing a check that the user has at least N of the collectionCards, not that every single one has been collected?

File: apps/jubmoji-quest/src/hooks/useFetchQuests.ts [hunk 2]
```diff
@@ -2,21 +2,24 @@ import { JubmojiQuest } from "@/types";
 import { useQuery } from "react-query";
 import { useJubmojis } from "./useJubmojis";
 import { $Enums, Prisma } from "@prisma/client";
+import { useFetchPowerById } from "./useFetchPowers";
 
-export const useGetQuestPowerLockedStatus = (questId?: string | number) => {
+export const useGetPowerLockedStatus = (
```
Comment (vb7401) [inline apps/jubmoji-quest/src/hooks/useFetchQuests.ts:7]: Should this code be moved to `useFetchPowers.ts`?

File: apps/jubmoji-quest/src/hooks/useFetchPowers.ts [hunk 3]
```diff
@@ -34,10 +34,12 @@ export const useFetchPowers = () => {
   );
 };
 
-export const useFetchPowerById = (id: string | number) => {
+export const useFetchPowerById = (id: string | number | undefined) => {
   return useQuery(
     ["powers", id],
     async (): Promise<JubmojiPower | null> => {
+      if (id === undefined) return null;
```
Comment (vb7401) [inline apps/jubmoji-quest/src/hooks/useFetchPowers.ts:41]: return null or undefined?

File: apps/jubmoji-quest/src/pages/quests/[id].tsx [hunk 4]
```diff
@@ -161,14 +157,24 @@ export default function QuestDetailPage() {
 
   const showLeaderboard = quest.proofType === $Enums.ProofType.TEAM_LEADERBOARD;
 
-  const collectionCardIndices = quest.collectionCards.map((card) => card.index);
-
-  const collectedItems =
-    jubmojis?.filter((jubmoji) =>
-      collectionCardIndices.includes(jubmoji.pubKeyIndex)
-    ).length ?? 0;
+  const collectionCardIndices = quest.powers
```
Comment (vb7401) [inline apps/jubmoji-quest/src/pages/quests/[id].tsx:160]: So quest collection cards are not taken into account here?

File: apps/jubmoji-quest/src/pages/quests/[id].tsx [hunk 5]
```diff
@@ -161,14 +157,24 @@ export default function QuestDetailPage() {
...
+    .flatMap((power) => power.collectionCards.map((card) => card.index))
+    .reduce((unique: number[], index: number) => {
+      if (!unique.includes(index)) {
+        unique.push(index);
+      }
+      return unique;
+    }, []);
 
-  const collectionTotalItems = quest.collectionCards.length;
+  const numPowersCompleted = quest.powers.filter((power) =>
```
Comment (vb7401) [inline apps/jubmoji-quest/src/pages/quests/[id].tsx:169]: Same comment here as before; need to check the power's rules before computing "numPowersCompleted"

File: apps/jubmoji-quest/src/pages/quests/[id].tsx [hunk 6]
```diff
@@ -243,10 +248,10 @@ export default function QuestDetailPage() {
                         {quest.proofType ===
                           $Enums.ProofType.TEAM_LEADERBOARD &&
```
Comment (vb7401) [inline apps/jubmoji-quest/src/pages/quests/[id].tsx:249]: We should probably have another quest DB column that determines whether the collection is hidden on frontend; this isn't actually tied to team leaderboard but was just a hack for the hunt

File: apps/jubmoji-quest/src/pages/quests/[id].tsx [hunk 7]
```diff
@@ -263,13 +268,39 @@ export default function QuestDetailPage() {
...
+            ).length ?? 0;
+
+          const proofParams = power.proofParams as Prisma.JsonObject;
+          let powerIsLocked: boolean;
+          let numCardsCollected: number;
+          let numCardsTotal: number;
+          if (power.proofType === $Enums.ProofType.N_UNIQUE_IN_COLLECTION) {
+            const N = proofParams.N as number;
+            powerIsLocked = collectedItems < N;
+            numCardsCollected = Math.max(collectedItems, N);
```
Comment (vb7401) [inline apps/jubmoji-quest/src/pages/quests/[id].tsx:286]: min?

File: apps/jubmoji-quest/src/pages/quests/[id].tsx [hunk 8]
```diff
@@ -263,13 +268,39 @@ export default function QuestDetailPage() {
...
+            numCardsTotal = N;
+          } else {
+            powerIsLocked = collectedItems === 0;
+            numCardsCollected = Math.max(collectedItems, 1);
```
Comment (vb7401) [inline apps/jubmoji-quest/src/pages/quests/[id].tsx:290]: min?

**Review by vb7401 — COMMENTED** [review]
Bunch of code fixes necessary before merging

File: apps/jubmoji-quest/src/pages/quests/[id].tsx (hunk 6 above)
Comment (AndrewCLu) [reply apps/jubmoji-quest/src/pages/quests/[id].tsx:249]: https://github.com/jubmoji/jubmoji.quest/issues/167

File: apps/jubmoji-quest/src/pages/quests/[id].tsx (hunk 4 above)
Comment (AndrewCLu) [reply apps/jubmoji-quest/src/pages/quests/[id].tsx:160]: No

**Review by vb7401 — APPROVED** [review]
Yay!

**Pull request merged**